pkg/virt-controller/watch/vm_initializer.go

```diff
@@ -0,0 +1,363 @@
+func NewVirtualMachineInitializer(vmPresetInformer cache.SharedIndexInformer, vmInitInformer cache.SharedIndexInformer, queue workqueue.RateLimitingInterface, vmInitCache cache.Store, clientset kubecli.KubevirtClient) *VirtualMachineInitializer {
+	vmi := VirtualMachineInitializer{
+		vmPresetInformer: vmPresetInformer,
+		vmInitInformer:   vmInitInformer,
+		clientset:        clientset,
+		queue:            queue,
+		store:            vmInitCache,
+	}
+	return &vmi
+}
+
+func (c *VirtualMachineInitializer) Run(threadiness int, stopCh chan struct{}) {
+	defer controller.HandlePanic()
+	defer c.queue.ShutDown()
+	log.Log.Info("Starting Virtual Machine Initializer.")
+
+	// Wait for cache sync before we start the pod controller
+	cache.WaitForCacheSync(stopCh, c.vmPresetInformer.HasSynced, c.vmInitInformer.HasSynced)
+
+	// Start the actual work
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(c.runWorker, time.Second, stopCh)
+	}
+
+	<-stopCh
+	log.Log.Info("Stopping controller.")
+}
+
+func (c *VirtualMachineInitializer) runWorker() {
+	for c.Execute() {
+	}
+}
+
+func (c *VirtualMachineInitializer) Execute() bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+	err := c.execute(key.(string))
+
+	if err != nil {
+		log.Log.Reason(err).Infof("reenqueuing VM %v", key)
+		c.queue.AddRateLimited(key)
+	} else {
+		log.Log.V(4).Infof("processed VM %v", key)
+		c.queue.Forget(key)
+	}
+	return true
+}
+
+func (c *VirtualMachineInitializer) execute(key string) error {
+
+	// Fetch the latest VM state from cache
+	obj, exists, err := c.store.GetByKey(key)
+
+	if err != nil {
+		return err
+	}
+
+	// If the VM isn't in the cache, it was just deleted, so shouldn't
+	// be initialized
+	if exists {
+		var vm *kubev1.VirtualMachine
+		vm = obj.(*kubev1.VirtualMachine)
+		// only process VM's that aren't initialized by this controller yet
+		if !isInitialized(vm) {
+			return c.initializeVirtualMachine(vm)
+		}
+	}
+
+	return nil
+}
+
+func (c *VirtualMachineInitializer) initializeVirtualMachine(vm *kubev1.VirtualMachine) error {
+	// All VM's must be marked as initialized or they are held in limbo forever
+	// Collect all errors and defer returning until after the update
+	logger := log.Log
+	errors := []error{}
+
+	logger.Object(vm).Info("Initializing VirtualMachine")
+
+	allPresets, err := listPresets(c.vmPresetInformer, vm.GetNamespace())
+
+	matchingPresets := filterPresets(allPresets, vm)
+
+	if len(matchingPresets) != 0 {
+		err = applyPresets(vm, matchingPresets)
+		if err != nil {
+			// A more specific error should have been logged during the applyPresets call.
+			// We don't know *which* preset in the list was problematic at this level.
+			logger.Object(vm).Errorf("Unable to apply presets to virtualmachine: %v", err)
+			errors = append(errors, err)
+		}
+	}
+
+	logger.Object(vm).Info("Marking VM as initialized and updating")
+	removeInitializer(vm)
+	_, err = c.clientset.VM(vm.Namespace).Update(vm)
```

Comment:
The key needs to be rate limited if any call to the k8s api server fails.

Reply:
@davidvossel could you clarify what you mean? It appears this was always the case that errors here caused a rate-limited retry.

Reply:
this looks correct. It's possible this was a comment that was related to the previous logic where errors were aggregated instead of resulting in an immediate 'return err'.